### PR TITLE
resolve #1727: Allow fallback after records already returned

### LINKED ISSUE #1727: Improve remote fetch fallback  to allow fallback after a few records have been returned
The fallback mode prevents fallback in the case records have been returned by the primary cursor.
This change allows the fallback cursor to resume iteration in the secondary cursor even if some records have already been returned.
This change requires that the continuations between the cursors is compatible

### PR DESCRIPTION
**Pull request by ohadzeliger**
Improve remote fetch fallback  to allow fallback after a few records have been returned

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java [hunk 1]
```diff
@@ -31,48 +31,57 @@
...
- * Note that there are business rules around when a fallback may happen. For example, because of the way records are
- * returned to the consumer, once a record is returned from the primary cursor, no fallback may be permitted: This is
- * done in order to prevent the case where a few records are returned, then a failure happens and the fallback cursor
- * starts again from the beginning, resulting in duplicate records being returned.
- * In practice, since many errors are observed when the request is sent to FDB (which coincide with the cursor's
- * first <code>onNext()</code> call, many such failures will be caught by that first result.
- *
+ * In order to continue from a failed iteration, the cursor will pass the last successful result to the provider
+ * of the fallback cursor. <b>It is the responsibility</b> of the fallback cursor provider to ensure that it can resume
+ * the iteration from that point. Among other things, that it can create an iteration over the range that starts with
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java:42]:
```suggestion
 * the iteration from that point. Among other things, this means starting a scan over the range that starts with
```

This is maybe a little clearer (arguably)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java [hunk 2]
```diff
@@ -31,48 +31,57 @@
...
+ * the failure point. If it can't, it can check whether the <code>lastSuccessfulResult</code> is <code>null</code> (in
+ * which case it can start from the beginning) and reject the case where the <code>inner</code> cursor has already
+ * returned records.
+ * <p></p>
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java:46]:
```suggestion
 * <p>
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/RemoteFetchTest.java [hunk 3]
```diff
@@ -292,15 +307,97 @@ void testReadYourWriteOutOfRangeSucceeds(RecordQueryPlannerConfiguration.IndexFe
...
+
+        List<TestRecords1Proto.MySimpleRecord> created = saveManyRecords();
+
+        // Modify record and then scan unmodified index
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context, splitRecordsHook);
+
+            // Update a record. This record will eventually be returned in a query, but we do *not* modify
+            // a field in the index being scanned, so that FDB does not detect the range conflict until later
+            TestRecords1Proto.MySimpleRecord lastRecord = created.get(created.size() - 1);
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/RemoteFetchTest.java:374]:
```suggestion
            TestRecords1Proto.MySimpleRecord lastRecord = created.get(0);
```
I think for the reverse test, it should edit the first result, as that lets it scan a bunch of results first before hitting the one in the RYW cache

Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/RemoteFetchTest.java:374]:
```suggestion
            TestRecords1Proto.MySimpleRecord lastRecord = created.get(0);
```
I think for the reverse test, it should edit the first result, as that lets it scan a bunch of results first before hitting the one in the RYW cache

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/RemoteFetchTest.java (hunk 3 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/RemoteFetchTest.java]: Absolutely, that was the intent. Fixed.

**Review by alecgrieser — DISMISSED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java [hunk 4]
```diff
@@ -31,48 +31,57 @@
...
- *
- * Note that there are business rules around when a fallback may happen. For example, because of the way records are
- * returned to the consumer, once a record is returned from the primary cursor, no fallback may be permitted: This is
- * done in order to prevent the case where a few records are returned, then a failure happens and the fallback cursor
- * starts again from the beginning, resulting in duplicate records being returned.
- * In practice, since many errors are observed when the request is sent to FDB (which coincide with the cursor's
- * first <code>onNext()</code> call, many such failures will be caught by that first result.
- *
+ * In order to continue from a failed iteration, the cursor will pass the last successful result to the provider
+ * of the fallback cursor. <b>It is the responsibility</b> of the fallback cursor provider to ensure that it can resume
```
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/FallbackCursor.java]: Done.

**Review by alecgrieser — DISMISSED** [review]

**Review by MMcM — APPROVED** [review]

**Review by alecgrieser — APPROVED** [review]

**Pull request merged**